### PR TITLE
feat: dynamic usernames

### PR DESCRIPTION
**Pull request by TJM**
Support using dynamic usernames with vault Username Templates.

### CODE REVIEW

**Comment by TJM** [comment]
As soon as I try to add the template, things get ugly, we are *very* far behind on the SDK version. If I can't make it work with the older version we are using, then I will have to upgrade all the other modules as well. 🤞

**Comment by alexhung** [comment]
> As soon as I try to add the template, things get ugly, we are _very_ far behind on the SDK version. If I can't make it work with the older version we are using, then I will have to upgrade all the other modules as well. 🤞

That is on my road map so if you get to this first, be my guess! 😄

**Comment by TJM** [comment]
Current POC functionality:


* DEFAULT, current functionality (after `make setup`):

```console
[tmcneely@local artifactory-secrets-plugin]$ vault read artifactory/token/test
Key                Value
---                -----
lease_id           artifactory/token/test/xFQp4bBihnOCVi0GCTq4vhVU
lease_duration     2h
lease_renewable    true
access_token       eyJ2ZXIiOiIyIiwidHlwIjoiSldUIiwiYWxnIjoiUlMyNTYiLCJraWQiOiJLMXptbEVTR2xpZlZXMnFtczdORmxpVENfVVMwYzVncTZBbzIyZUc5UHhjIn0.eyJzdWIiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxXC91c2Vyc1wvdGVzdC11c2VyIiwic2NwIjoiYXBwbGllZC1wZXJtaXNzaW9uc1wvZ3JvdXBzOnJlYWRlcnMiLCJhdWQiOiIqQCoiLCJpc3MiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxIiwiaWF0IjoxNjc4MzEyNTUzLCJqdGkiOiIyNTBhNzdjMC1jMTVkLTQwMmQtOWIxNi05YzA2ODE5ZTk5ZmQifQ.LCdktZ3gK1Wd1IB19lL3zi9HIwN7Y57eQsqxknoQX3iLWiam_Vrlk-7acIkCLqrvkJy40AZz7Kxp8foaeZvWqpVgdymji8KzYb9NTr_8T6KARlzfVVb_p_TQ77adS5ZL0sXABXbgTMBbQ4a2n0WtkMM7z3h4KJuVDjSN7Xbxcs_Qur7s82EqGThPQM6zv28KlXSU5mwEe7Qnu23Us5Og3U6x7p0p9Znr9GpnFq3jV_4TugYC1Yj2hPT2addQ_08OwvjdtkkNbJIMxG3eXF0V3yv8GlzK1aR1xeJsNXTxqOtU8FUnlt-PfefjVOehv_y0kvKNkhHpxo3_JdeJlm8LSg
role               test
scope              applied-permissions/groups:readers
token_id           250a77c0-c15d-402d-9b16-9c06819e99fd
username           test-user
```

* UNSET the static username for the role (same as not setting it in the first place:

```console
[tmcneely@local artifactory-secrets-plugin]$ vault write artifactory/roles/test username=""
Success! Data written to: artifactory/roles/test
```


* New operation, generate a dynamic username (for now its a static string, for PoC):

```console
[tmcneely@local artifactory-secrets-plugin]$ vault read artifactory/token/test            
Key                Value
---                -----
lease_id           artifactory/token/test/XgQHjbIRzV0tPjNYETqF1dpK
lease_duration     2h
lease_renewable    true
access_token       eyJ2ZXIiOiIyIiwidHlwIjoiSldUIiwiYWxnIjoiUlMyNTYiLCJraWQiOiJLMXptbEVTR2xpZlZXMnFtczdORmxpVENfVVMwYzVncTZBbzIyZUc5UHhjIn0.eyJzdWIiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxXC91c2Vyc1wvZnV0dXJlLXRlbXBsYXRlLXVzZXIiLCJzY3AiOiJhcHBsaWVkLXBlcm1pc3Npb25zXC9ncm91cHM6cmVhZGVycyIsImF1ZCI6IipAKiIsImlzcyI6ImpmYWNAMDFndHl6N3IybjJjcW0wYXp3NHdzaDAxcnEiLCJpYXQiOjE2NzgzMTI2MDYsImp0aSI6ImI5MmZmZWUwLTk4NjAtNDM5NS1hYTAyLWRlNzc0ZGYzMmJmYyJ9.iDKt1JFD2r6x6soHlPknq_I-uaXUYr1hYPVT0_6tPAcz0d_1x-yTnIopX4ModlhL9miulgKG9ShohR2No4BU4VA52t6oJtpkQ1-Oy_NEykac1HShwIi8cPLUUn4uX3tRJaKenyvNvILqh6yMlp2_jdQ350c9McUav_LszgcGOTXZEVpfmTQIie4sO1IaIyjtt9ZTCY0rdyT_xu9jePqk7kSbGA2UyhE2wYS2lOprC5R75cZkUu2Pvmgf_phlLF3xdkKInUPJzPvIXnlvLu353CoJ1-XmW8j9xqY2wkR_gncOyVeWUAq9SnahGpPzGXGi7xP2acjpb3GxsU8S294AjA
role               test
scope              applied-permissions/groups:readers
token_id           b92ffee0-9860-4395-aa02-de774df32bfc
username           future-template-user
```

**Comment by TJM** [comment]
New version:

```console
[tmcneely@local artifactory-secrets-plugin]$ vault read artifactory/token/test             
Key                Value
---                -----
lease_id           artifactory/token/test/ts3UvVwJpIqS2f7Sx5EZpa0q
lease_duration     2h
lease_renewable    true
access_token       eyJ2ZXIiOiIyIiwidHlwIjoiSldUIiwiYWxnIjoiUlMyNTYiLCJraWQiOiJLMXptbEVTR2xpZlZXMnFtczdORmxpVENfVVMwYzVncTZBbzIyZUc5UHhjIn0.eyJzdWIiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxXC91c2Vyc1wvdi10ZXN0LXkybWdrcGxyY28yaHVtY3V5d3B5Iiwic2NwIjoiYXBwbGllZC1wZXJtaXNzaW9uc1wvZ3JvdXBzOnJlYWRlcnMiLCJhdWQiOiIqQCoiLCJpc3MiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxIiwiaWF0IjoxNjc4MzE5MTMzLCJqdGkiOiJiYzFlMmI1YS03ZjAyLTRlMmMtYTIwOC03MGI4NzVjZjFkODAifQ.eZpJ3yb2jtp94ey2O9biLeS-BuWvvakYY8N0YIqCJLVdthqgOKVzt0vXisHi1k65hUyv2w1ZG74iJA4HG6yu8ph5cqUlcq9hHAX0WpjSWhPpIOcKs16FiyAF2GvSr5MbfIMa_xZsuQ2UOtWUCNq8U4OEym9ktf2FJbEcs8atEkxpkSUSD7T8GuVQiToDyI_kDbaPqzVLbyYpzHZJzV4sIa4ITdOmPRhsBaTnjecKRy1C5Hc5gdyotH_FHdGpDmSEPMoXwi5z7YVDPFVD1gyhxlRqWo-uUj9H56f-wBqbPlVuOgXKCDninPVgKQqEY5mvtfDe3ifTzwhUSIDqL8eFUA
role               test
scope              applied-permissions/groups:readers
token_id           bc1e2b5a-7f02-4e2c-a208-70b875cf1d80
username           v-test-y2mgkPlrco2humCUYWPY
```

Also, customizable templates:

```console
[tmcneely@local artifactory-secrets-plugin]$ vault write artifactory/config/admin username_template='{{printf "v_%s_%s_%s_%s" (.DisplayName | truncate 8) (.RoleName | truncate 8) (random 20) (unix_time) | truncate 256}}'
Success! Data written to: artifactory/config/admin
[tmcneely@local artifactory-secrets-plugin]$ vault read artifactory/token/test
Key                Value
---                -----
lease_id           artifactory/token/test/XOhLnCo7BqeIuo3FfVeoABwh
lease_duration     2h
lease_renewable    true
access_token       eyJ2ZXIiOiIyIiwidHlwIjoiSldUIiwiYWxnIjoiUlMyNTYiLCJraWQiOiJLMXptbEVTR2xpZlZXMnFtczdORmxpVENfVVMwYzVncTZBbzIyZUc5UHhjIn0.eyJzdWIiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxXC91c2Vyc1wvdl9fdGVzdF9tMHF3dTk1ZWx5c3dmdWQ2bDY0el8xNjc4MzE5MjczIiwic2NwIjoiYXBwbGllZC1wZXJtaXNzaW9uc1wvZ3JvdXBzOnJlYWRlcnMiLCJhdWQiOiIqQCoiLCJpc3MiOiJqZmFjQDAxZ3R5ejdyMm4yY3FtMGF6dzR3c2gwMXJxIiwiaWF0IjoxNjc4MzE5MjczLCJqdGkiOiJmOGM3ZDU1Ni00OTViLTRkM2YtODlkYS0yYTk2YzU5ZDQwMmUifQ.MYkJzsQixPdLan63KHxb0uBN4zukYC6CXxxSbcoxc2w-OO9wLIfnjNVSfcn8OEjO03lzHbnaScjtXbZtKqBuTjwv4gbQUHNqwf-utWX7mGard-47QAkv2Th_ljTV1otpeGMerbWLnxwVmgZSxJGf9mR9W8xfQo0H-F25ENd35-ctpyjR3w1OCCpJd5bHXKnUFjFEYozT9KwjjsSMP75y341ySKh8uIYrezbvpcqRyT9grEFA0-vS2vAS_5Uv765yNPka5O7QPaiJM0phgAyeGpsFFNAkcIpfVcyf1nEfgy6XKvPdKHvpsYq748YCr7jjlGUrufUxPAXMk9MSfttU_Q
role               test
scope              applied-permissions/groups:readers
token_id           f8c7d556-495b-4d3f-89da-2a96c59d402e
username           v__test_M0Qwu95ELYSWFuD6l64Z_1678319273
[tmcneely@local artifactory-secrets-plugin]$ 
```

I am still not sure what to put for some of the fields, nor whether it will survive a reboot (99% sure it wont). So, still in draft, but getting closer to "MVP" :)

**Comment by alexhung** [comment]
@TJM I'm assuming that those `access_token` are not valid ones but you may want to sanitize them before posting here 😄

**Comment by TJM** [comment]
> @TJM I'm assuming that those `access_token` are not valid ones but you may want to sanitize them before posting here 😄

They are for my localhost artifactory-jcr. I wanted to leave them in to show how artifactory itself handled the request.

File: backend.go [hunk 1]
```diff
@@ -40,6 +48,12 @@ func Backend(_ *logical.BackendConfig) (*backend, error) {
 		httpClient: http.DefaultClient,
 	}
 
+	up, err := testUsernameTemplate(defaultUserNameTemplate)
```
Comment (TJM) [inline backend.go:51]: This works fine for initialization, but I think we may need to do some "work" to handle a restart. I will have to setup a proper vault to test in, cause dev mode of course wipes everything out when it restarts.

File: path_config.go [hunk 2]
```diff
@@ -47,25 +51,49 @@ individual calls, so do not include it in the URL here.
...
+	if config == nil {
+		config = &adminConfiguration{}
+	}
+
+	if val, ok := data.GetOk("access_token"); ok {
+		config.AccessToken = val.(string)
+	}
+	if val, ok := data.GetOk("url"); ok {
+		config.ArtifactoryURL = val.(string)
+	}
```
Comment (TJM) [inline path_config.go:88]: I have to think there is a smarter way to do this, but this allows you to change the username_template without setting the url and token every time. We *might* want to require the token if the URL is set... that would be a way to expose the token by pointing to a webserver you had control of.

File: path_config.go [hunk 3]
```diff
@@ -120,12 +148,17 @@ func (b *backend) pathConfigRead(ctx context.Context, req *logical.Request, _ *f
...
+		configMap["username_template"] = config.UsernameTemplate
+	}
+
 	// Optionally include token info if it parses properly
 	token, err := b.getTokenInfo(*config, config.AccessToken)
 	if err != nil {
 		b.Logger().Warn("Error parsing AccessToken: " + err.Error())
 	} else {
-		configMap["jfrog_token_id"] = token["TokenID"]
+		configMap["token_id"] = token["TokenID"]
```
Comment (TJM) [inline path_config.go:161]: shortened this to `token_id` for consistency.

File: path_roles.go [hunk 4]
```diff
@@ -35,8 +35,7 @@ func (b *backend) pathRoles() *framework.Path {
 			},
 			"username": {
 				Type:        framework.TypeString,
-				Required:    true,
-				Description: `Required. The username for which the access token is created. If the user does not exist, Artifactory will create a transient user. Note that non-admininstrative access tokens can only create tokens for themselves.`,
+				Description: `Optional. Defaults to using the username_template. The static username for which the access token is created. If the user does not exist, Artifactory will create a transient user. Note that non-administrative access tokens can only create tokens for themselves.`,
```
Comment (TJM) [inline path_roles.go:38]: This help text might be a bit wordy. I don't think we need the information about admin tokens here.

File: path_roles_test.go [hunk 5]
```diff
@@ -30,7 +30,6 @@ func TestBackend_PathRoleList_CannotAddRoleWhenNotConfigured(t *testing.T) {
 	b, config := makeBackend(t)
 
 	roleData := map[string]interface{}{
-		"role":     "test-role",
```
Comment (TJM) [inline path_roles_test.go:33]: I had to remove this, as in the updated version of (something), it sent a warning about the role being used from path data instead of the parameter, which caused the test to fail.

File: path_token_create.go [hunk 6]
```diff
@@ -75,6 +75,17 @@ func (b *backend) pathTokenCreatePerform(ctx context.Context, req *logical.Reque
 		return logical.ErrorResponse("no such role"), nil
 	}
 
+	// Define username for token by template if a static one is not set
+	if len(role.Username) == 0 {
+		role.Username, err = b.usernameProducer.Generate(UsernameMetadata{
+			RoleName: roleName,
+			// DisplayName: "Not Sure",
```
Comment (TJM) [inline path_token_create.go:82]: AS with the UsernameMetadata field above, we have to decide what these should be set to, and what other fields might be useful when generating a templated username.

File: path_roles_test.go (hunk 5 above)
Comment (TJM) [reply path_roles_test.go:33]: Role is a path parameter. I wonder if this test was just from a different time. :)

File: path_config.go (hunk 2 above)
Comment (TJM) [reply path_config.go]: Resolved, move the check/set for `access_token` below `url` and if url is set, it clears the access token, so it won't expose the current access_token to a new url.

File: artifactory.go [hunk 7]
```diff
@@ -16,9 +16,14 @@ import (
 
 	jwt "github.com/golang-jwt/jwt/v4"
 	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/vault/sdk/helper/template"
 	"github.com/hashicorp/vault/sdk/logical"
 )
 
+const (
+	defaultUserNameTemplate string = `{{ printf "v-%s-%s" (.RoleName | truncate 15) (random 20) | truncate 50 }}`
```
[the remark this answers is not in the document]
Comment (TJM) [reply artifactory.go]: Changed to random 8, also allowed more characters from the RoleName.

File: Makefile [hunk 8]
```diff
@@ -41,11 +41,18 @@ clean:
...
 	go fmt $$(go list ./...)
 
-setup: disable enable
+setup: disable enable admin testrole
+
+admin:
 	vault write artifactory/config/admin url=$(ARTIFACTORY_URL) access_token=$(JFROG_ACCESS_TOKEN)
 	vault read artifactory/config/admin
-	vault write artifactory/roles/test scope="$(ARTIFACTORY_SCOPE)" username="test-user" max_ttl=3h default_ttl=2h
+	vault write -f artifactory/config/rotate
```
Comment (TJM) [inline Makefile:49]: NOTE: Also added a token rotate here, because the "script" generated token (at least on v7.55.2) is only valid for 1h. This `make setup` is almost a full functional test.

File: backend.go (hunk 1 above)
Comment (TJM) [reply backend.go:51]: I am having issues running in "not -dev" mode... it acts like it is failing the mTLS init, even when the vault server has no TLS. I have tried different version of the plugin and different versions of vault. No luck. :(

Comment (alexhung) [reply backend.go:51]: I assume you have set [`tls_disable: true`](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp#tls_disable)? Or the `tls_disable_client_certs`?

Comment (alexhung) [reply backend.go:51]: Or set `-tls-skip-verify` in the cli?

Comment (TJM) [reply backend.go:51]: Tried both with and without TLS. I was able to reproduce the same concept by using `vault plugin reload -plugin artifactory` ... and the version stuff disappeared, so I am going to start looking at initialize now :)

> https://github.com/hashicorp/vault/discussions/19509

File: path_token_create.go (hunk 6 above)
Comment (TJM) [reply path_token_create.go:82]: @alexhung - Let me know if you have any ideas what to put here, I think this is the "Display Name" of the authentication token in database land? We can also add other fields.

File: path_roles.go (hunk 4 above)
Comment (TJM) [reply path_roles.go:38]: @alexhung - Should we trim it down or leave it alone?

File: backend.go (hunk 1 above)
Comment (TJM) [reply backend.go:52]: Lets not let my local issues cause delays here, this is done.

File: backend.go [hunk 9]
```diff
@@ -8,14 +8,22 @@ import (
...
+	rolesMutex       sync.RWMutex
+	httpClient       *http.Client
+	usernameProducer template.StringTemplate
+}
+
+// UsernameMetadata defines the metadata that a user_template can use to dynamically create user account in Artifactory
+type UsernameMetadata struct {
+	DisplayName string
+	RoleName    string
 }
```
[the remark this answers is not in the document]
Comment (TJM) [reply backend.go:28]: Closing this in favor of the comments below in `path_create_token.go`

File: path_token_create.go (hunk 6 above)
Comment (alexhung) [reply path_token_create.go:82]: May be the description of the token? That field is not part of the code base currently but would be good to add (separate PR?).

File: path_roles.go (hunk 4 above)
Comment (alexhung) [reply path_roles.go:38]: I tend to err towards verbose description. Alternatively, include a link to the relevant documentation. One of the section in this https://www.jfrog.com/confluence/display/JFROG/Access+Tokens?

File: path_token_create.go (hunk 6 above)
Comment (TJM) [reply path_token_create.go:82]: @davidcorrigan714 - Your feedback would be nice too. If you were generating a dynamic username in artifactory for a CI/CD job or something, what type of information would be good to pass to the template engine?

File: path_roles.go (hunk 4 above)
Comment (TJM) [reply path_roles.go:38]: This is help (if anyone can figure out the `path-help` vault command), for the username for the role. It seems like the wrong place.

File: path_token_create.go (hunk 6 above)
Comment (davidcorrigan714) [reply path_token_create.go:82]: Not too familiar with the templates, but if DisplayName is the authenticated user to Vault then that looks to be pretty useful and seems to be a common field used in the other plugins that support this.

**Review by alexhung — APPROVED** [review]
LGTM

And thanks so much for improving the README!

The only outstanding item seems to be for the `UsernameMetadata.DisplayName`.

File: path_token_create.go (hunk 6 above)
Comment (TJM) [reply path_token_create.go:82]: I think displayName might be useful in the description (which will be another MR), but I am not sure that I want "Firstname Lastname" embedded into the username for artifactory. (spaces in usernames scare me)

Tommy

File: backend.go (hunk 1 above)
Comment (TJM) [reply backend.go:52]: I upgraded the API version to 1.9.0, which solved the problems I was seeing in our "real" Vault instance. RESOLVED!

**Comment by TJM** [comment]
FYI, depending on the order these are merged, the other may have to be manually rebased, since they touch some of the same files.

**Comment by alexhung** [comment]
The other PR is already merged. So you should resolve merge conflict in this branch/PR.

**Comment by TJM** [comment]
Rebased, had to change the `b.suppportForceRevocable` to use the new `b.checkVersion` (with the saved version), as expected.

**Comment by TJM** [comment]
Fixes #46

**Comment by alexhung** [comment]
@TJM Let @shrajfr12 and I know when this PR is ready for review again.

**Comment by TJM** [comment]
@alexhung / @shrajfr12 - Should be good to go. I ran a quick test with `make setup` but I think testing is something we need to dig deeper into ;)

**Review by alexhung — APPROVED** [review]

**Review by shrajfr12 — APPROVED** [review]
lgtm
